src/Features/CSharp/Portable/SignatureHelp/AbstractCSharpSignatureHelpProvider.cs

```diff
@@ -72,5 +80,288 @@ protected IList<TaggedText> GetAwaitableUsage(IMethodSymbol method, SemanticMode
 
             return SpecializedCollections.EmptyList<TaggedText>();
         }
+
+        /// <summary>
+        /// If the symbol could not be bound, we could be dealing with a partial invocation, we'll try to find a possible overload.
+        /// If it was bound, we'll find which parameter to highlight.
+        /// Either way, we'll eliminate unsuitable candidates (based on argument names used in the invocation).
+        /// </summary>
+        protected static (ISymbol symbol, int parameterIndex) GuessCurrentSymbolAndParameter(
+            SeparatedSyntaxList<ArgumentSyntax> arguments, ImmutableArray<IMethodSymbol> methodGroup, int position,
+            SemanticModel semanticModel, ISemanticFactsService semanticFactsService, CancellationToken cancellationToken)
+        {
+            if (arguments.Count != 0)
+            {
+                foreach (var method in methodGroup)
+                {
+                    if (IsAcceptable(arguments, method, position, semanticModel, semanticFactsService, out var parameterIndex))
+                    {
+                        return (method, parameterIndex);
+                    }
+                }
+            }
+
+            // Note: Providing no recommendation if no arguments allows the model to keep the last implicit choice
+            return (null, -1);
+        }
+
+        /// <summary>
+        /// If an argument name does not correspond to a parameter name, this method is unacceptable. We'll filter it out.
+        /// </summary>
+        protected static bool IsUnacceptable(SeparatedSyntaxList<ArgumentSyntax> arguments, IMethodSymbol method)
+        {
+            if (arguments.Count == 0)
+            {
+                return false;
+            }
+
+            var parameterNames = method.Parameters.SelectAsArray(p => p.Name);
+            foreach (var argument in arguments)
+            {
+                if (HasName(argument, out var name))
+                {
+                    if (!parameterNames.Contains(name))
+                    {
+                        return true;
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        protected static bool IsAcceptable(SeparatedSyntaxList<ArgumentSyntax> arguments, IMethodSymbol method, int position,
+            SemanticModel semanticModel, ISemanticFactsService semanticFactsService, out int foundParameterIndex)
```

Comment:
having IsAceptable and IsUnacceptable, and having them be non-symettric is freaking me out.

Reply:
might want to doc what foundParameterIndex is.

Reply:
this feels like a better name would be "TryFindParameterIndex" given the bool+outvar

Reply:
can you give a doc comment on this method?  it's complex enough that i don't understand the overall purpose.

Reply:
I'll add comments.

IsUnacceptable filters methods so they are not in the output. For example, if you use a named argument that doesn't exist in some overloads.
IsAcceptable simulates overload resolution and tells you if this overload is compatible with the arguments you're passing. If the overload is acceptable, then it also returns an index that tells you which parameter should be highlighted in sighelp.

Reply:
TryFindCompatibleMethodAndMatchingParameterIndex?   Wordy, but pretty clear to me :)

Reply:
In explaining this, I realized two things: (1) the tests don't really verify the highlighted parameters, and (2) the highlighted parameters are probably wrong when named arguments are involved and the parameter names are in different orders in different overloads.

This highlight behavior is probably no worse than our shipped behavior, and it should be better for common cases. But it's probably worth taking more time to do this right. I'll move this PR out to 16.1 preview1.

Reply:
After investigating my concern further:

It used to be we can just compute argumentIndex and argumentName (I think name supercedes index if present) for the whole set of results. (see `SignatureHelpTestItems`)
If you started typing a new argument after a named argument, we highlight an incorrect parameter (just based on number of arguments you typed so far).

But now, I do something better for that scenario (finding the first unspecified argument). But that is only applicable for the selected item, but it is still incorrect for the rest of the results.

I think it's not worthwhile to try to fix that at the moment, especially given this is an existing problem.

Note: For common scenarios (you're typing an in-position argument, or a named argument), the highlight is still good for the whole set of results.

What do you think?